cli/fathom_web/commands/vectorize.py

```diff
@@ -0,0 +1,337 @@
+    return str(xpi_path.absolute())
+
+
+def run_file_server(samples_directory):
+    """
+    Create a local HTTP server for the samples.
+
+    We return both the server and the thread to handle a graceful or forced
+    shutdown.
+    """
+    print('Starting HTTP file server...', end='', flush=True)
+    RequestHandler = partial(SilentRequestHandler, directory=samples_directory)
+    server = HTTPServer(('localhost', 8000), RequestHandler)
+    server_thread = Thread(target=server.serve_forever)
+    server_thread.start()
+    print('Done')
+    return server, server_thread
+
+
+def configure_firefox(fathom_fox, fathom_trainees, output_directory, show_browser, temp_dir):
+    """
+    Configures and launches Firefox to run the vectorizer with.
+
+    Sets headless mode, sets the download directory to the desired output
+    directory, turns off page caching, and installs FathomFox and Fathom
+    Trainees.
+
+    We return the webdriver object, and the process IDs for both the Firefox
+    process and the geckodriver process so we can shutdown either
+    gracefully or ungracefully.
+    """
+    print('Configuring Firefox...', end='', flush=True)
+    options = webdriver.FirefoxOptions()
+    options.headless = not show_browser
+
+    profile = webdriver.FirefoxProfile()
+    profile.set_preference('browser.download.folderList', 2)
+    profile.set_preference('browser.download.dir', str(pathlib.Path(output_directory).absolute()))
+    profile.set_preference('browser.cache.disk.enable', False)
+    profile.set_preference('browser.cache.memory.enable', False)
+    profile.set_preference('browser.cache.offline.enable', False)
+
+    firefox = webdriver.Firefox(
+        options=options,
+        firefox_profile=profile,
+        service_log_path=f'{temp_dir.name}/geckodriver.log',
+    )
+
+    firefox.install_addon(fathom_fox, temporary=True)
+    firefox.install_addon(fathom_trainees, temporary=True)
+    print('Done')
+    return firefox, firefox.capabilities['moz:processID'], firefox.service.process.pid
+
+
+def run_vectorizer(firefox, fathom_type, sample_filenames):
+    """
+    Set up the vectorizer and run it, creating the vectors file.
+
+    Navigate to the vectorizer page of FathomFox, paste the sample filenames
+    into the text area and hit the vectorize button.
+
+    We monitor the status text area for errors and to see how many samples
+    have been vectorized, so we know when the vectorizer has stopped running.
+    """
+    print('Configuring Vectorizer...', end='', flush=True)
+    # Navigate to the vectorizer page
+    fathom_fox_uuid = get_fathom_fox_uuid(firefox)
+    firefox.get(f'moz-extension://{fathom_fox_uuid}/pages/vector.html')
+
+    pages_text_area = firefox.find_element_by_id('pages')
+    pages_text_area.send_keys(sample_filenames)
+
+    ruleset_dropdown_selector = Select(firefox.find_element_by_id('ruleset'))
+    ruleset_dropdown_selector.select_by_visible_text(fathom_type)
```

Comment:
When, for whatever reason FathomFox doesn’t find a `trainees.js`, it throws an [error](https://paste.mozilla.org/kiPcqCSS) of the form: `selenium.common.exceptions.NoSuchElementException: Message: Could not locate element with visible text: article`.

We should make this error message more helpful if possible as to the true issue, as I only learned it by using the `-s` option and briefly seeing the warning on the Vectorizer page before Firefox shut down ungracefully. This can be a follow-up issue.

Reply:
I agree.